Implement delete_before_upload feature

lveyde opened the pull request:
In some cases it may be desirable to make sure that uploading always creates new files, even if the the file(s) with the same names already exist in the target directory.

Otherwise if somebody hard-links to the existing file, the upload using the same filename will modify all the hard-linked files, which may come as a surprise and undesired behavior.

When using the delete_before_upload option (specified and set to yes) we'll first attempt to delete the file in the target directory, and only then to upload it there. Thus implementing Copy-On-Write in case one hard-links to the uploaded files.

lveyde committed:
Added an option to delete files before the upload

In some cases it may be important to make sure the files we upload won't reuse the same inode, i.e. in case the file(s) with the same name already exists in the target directory and have hard links.

Currently uploading a new file with the same name will silently push a change to all linked instances, something we may want to avoid.

Deleting the files before the upload takes care of this.

Signed-off-by: Lev Veyde <lveyde@redhat.com>

lveyde committed:
Added a test for delete_before_upload functionality

Signed-off-by: Lev Veyde <lveyde@redhat.com>

janosdebugs reviewed (changes requested):
Looks good, but the readme change is missing.

michalskrivanek commented:
lgtm too, but where do we have hardlinks, and why?

lveyde commented:
> lgtm too, but where do we have hardlinks, and why?

We normally hard-link(ed) to the existing packages in order to save disk space.
That is for example a part of the standard procedure during the preparation of the oVirt release as well.

Previously, when the package(s) were added manually that wasn't an issue, since in case the same filename was re-used, we were well aware of it, and could delete the file we linked to before updating it.

Now, with the uploading happens automatically, that is no longer the case. So it basically presents us with 3 options - first one is to always make a full copy of the files, some of which are over 1 GB in size, second is to (try to) carefully watch the cases where the files may have the same name, or the third one - which this patch implements, is to break the chain and always make the updated file to be a new one, unlinked to the previous version of file with the same name.

That case may require a manual re-linking to the newly uploaded file, but that is a conscious operation one does, not something which happens as a surprise.

michalskrivanek commented:
or symlinks?

currently this action is used for CI artifacts, not for a release. where do you intend to use it?

lveyde committed:
Updated the README.md for delete_before_upload functionality

Signed-off-by: Lev Veyde <lveyde@redhat.com>

lveyde commented:
> 
I planned to use it in https://github.com/oVirt/ovirt-node-ng-image/ i.e. https://github.com/oVirt/ovirt-node-ng-image/blob/master/.github/workflows/build.yml

However since it currently uses too wild wildcards I'll need to modify the code a bit, so seems that the patch needs a bit more work to function properly in this case.

lveyde commented:
> or symlinks?
> 
> currently this action is used for CI artifacts, not for a release. where do you intend to use it?

We don't use symlinks, and there is not much we (or anyone) can potentially do in case a symlink is used as it has no connection to the file, so to detect link to a specific file one has to traverse all filesystems (as symlinks can point to a file/directory on another filesystem) to check for a link.

lveyde committed:
Implemented source side wildcard (globbing) processing

Previously, if the source included any wildcards, these were processed at the target side, thus
potentially deleting more files that were later uploaded, as the same wildcard selection could
be expanded to a different set of files.

This patch modified the code to perform the expansion at the source side, thus only deleting
the same list of files that are later will be uploaded.

It also updates the test and README.md file for the change.

Signed-off-by: Lev Veyde <lveyde@redhat.com>

lveyde commented:
> > 
> 
> I planned to use it in https://github.com/oVirt/ovirt-node-ng-image/ i.e. https://github.com/oVirt/ovirt-node-ng-image/blob/master/.github/workflows/build.yml
> 
> However since it currently uses too wild wildcards I'll need to modify the code a bit, so seems that the patch needs a bit more work to function properly in this case.

OK, fixed that issue. Now it should work for our case as expected.

michalskrivanek commented:
sure, but i mean - why do you even use hardlink for anything? The only usecase I can think of here is to have links like "latest" pointing to certain uploaded artifacts. You have to change those in some different way anyway.

lveyde commented:
> sure, but i mean - why do you even use hardlink for anything? The only usecase I can think of here is to have links like "latest" pointing to certain uploaded artifacts. You have to change those in some different way anyway.

We use them to save disk space. That way we can point to a file without wasting a space on a duplicate, yet if the initial file is removed for whatever reason, we still have the file at any other locations. That way the same file i.e. large RPM package or ISO can appear in multiple places with near zero cost of the disk space, and we're letting the OS to handle the reference counting and make sure to delete the file only when it's really not pointed at anymore.

The only downside to this process (in our case) is that a change through any hard-links will modify the same file. As explained, this normally didn't happen when were adding the files manually, but with the introduction of the automatic uploading it may (and already did) happen.

So yes - we can instead waste disk space and always make a regular copy of the file(s) each time, or at very least from the locations where they're being uploaded to automatically, but IMHO fixing this issue is a much better solution.

michalskrivanek commented:
but it's not usable with this action, it can't create hardlinks. So how do you create them? And what's the difference/problem with symlinks? All the content on resources.ovirt.org is anyway just served by apache, it shouldn't matter at all if it's a symlink or not.

janosdebugs commented on action.yml:
```diff
@@ -139,6 +149,28 @@ runs:
+        echo "SOURCE is set to: ${SOURCE}"
+        FILES=$(echo $SOURCE)
+        echo "FILES is set to: ${FILES}"
+        COMMAND='set -e; cd '${TARGET}' && for file in '${FILES}'; do echo "file: $file" && rm -v $(basename $file) || true; done'
```
Is ${TARGET} going to be properly replaced this way?

lveyde replied:
> Is ${TARGET} going to be properly replaced this way?

Yes, since it's outside of single quoted string, just like ${FILES}:

Running remote command: set -e; cd /*** && for file in ***data/1 ***data/11 ***data/14 ***data/16 ***data/18 ***data/19; do echo "file: $file" && rm -v $(basename $file) || true; done
file: ***data/1
removed '1'
file: ***data/11
removed '11'
file: ***data/14
removed '14'
file: ***data/16
removed '16'
file: ***data/18
removed '18'
file: ***data/19
removed '19'

lveyde commented:
> but it's not usable with this action, it can't create hardlinks. So how do you create them? And what's the difference/problem with symlinks? All the content on resources.ovirt.org is anyway just served by apache, it shouldn't matter at all if it's a symlink or not.



> but it's not usable with this action, it can't create hardlinks. So how do you create them? And what's the difference/problem with symlinks? All the content on resources.ovirt.org is anyway just served by apache, it shouldn't matter at all if it's a symlink or not.

It's not supposed to create hard-links, we do that afterwards, manually, when we need to point to some of the uploaded files.

There are a few problems with symlinks:

- There is no an easy way to track them, i.e. there is no easy way to get all symlinks that point at a specific file without traversing all file-systems and resolve all links.
- If the file it's pointing to is gone, then the symlink becomes broken. For example if we point to a node-ng build, and later on the directory gets cleaned up, removing it for being too old. Then suddenly the repo is broken. With hard-links that won't happen, so less chances for nasty surprises.
- They're "stupid", in the way that they can't catch any change in the target they're pointing at, and simply always point to the same path. If you delete the target file, and then re-create a new one, the symlink will silently point to it like nothing happened. That is unlike the hard-link, which will disconnect the relation between the 2 (or more) files the instant it's being deleted, and won't re-create that relation even if the file with the same name is created at the same place - unless we specifically create it as a hard link of that same file.

We do use symlinks, in cases where we want to create meta pointers to the latest version, when we know that the target won't suddenly disappear (i.e. it's pointed at the specific ovirt-release version RPM inside of the repo), and using a symlink provides an easy way to see what version we're pointing at. Yet we have special software that we run as part of the release process that verifies that the links are not broken and point to the correct type of the package.
